gdx/src/com/badlogic/gdx/graphics/g3d/Environment.java

```diff
@@ -28,11 +28,95 @@
 import com.badlogic.gdx.utils.GdxRuntimeException;
 
 public class Environment extends Attributes {
+	
+	public static interface EnvironmentListener {
+		public void onLightAdded(PointLight pointLight);
+		public void onLightAdded(DirectionalLight directionalLight);
+		public void onLightAdded(SpotLight spotLight);
+		
+		public void onLightRemoved(PointLight pointLight);
+		public void onLightRemoved(DirectionalLight directionalLight);
+		public void onLightRemoved(SpotLight spotLight);
+	}
+	
+	private Array<EnvironmentListener> listeners = new Array<EnvironmentListener>();
+	
 	/** @deprecated Experimental, likely to change, do not use! */
+	@Deprecated
 	public ShadowMap shadowMap;
 
 	public Environment () {
 	}
+	
+	public Environment addListener(EnvironmentListener listener) {
+		listeners.add(listener);
+		
+		// Add all the existing lights
+		DirectionalLightsAttribute dirLights = ((DirectionalLightsAttribute)get(DirectionalLightsAttribute.Type));
+		PointLightsAttribute pointLights = ((PointLightsAttribute)get(PointLightsAttribute.Type));
+		SpotLightsAttribute spotLights = ((SpotLightsAttribute)get(SpotLightsAttribute.Type));
+		
+		if( dirLights != null ) {
+			for( DirectionalLight dirLight: dirLights.lights ) {
+				listener.onLightAdded(dirLight);
+			}
+		}
+		
+		if( pointLights != null ) {
+			for( PointLight pointLight: pointLights.lights ) {
+				listener.onLightAdded(pointLight);
+			}
+		}
+		
+		if( spotLights != null ) {
+			for( SpotLight spotLight: spotLights.lights ) {
+				listener.onLightAdded(spotLight);
+			}
+		}
+		
+		return this;
+	}
+	
+	private void addBroadcast(DirectionalLight dir) {
+		for(EnvironmentListener listener : listeners) {
+			listener.onLightAdded(dir);
+		}
+	}
+	
+	private void addBroadcast(PointLight point) {
+		for(EnvironmentListener listener : listeners) {
+			listener.onLightAdded(point);
+		}
+	}
+	
+	private void addBroadcast(SpotLight spot) {
+		for(EnvironmentListener listener : listeners) {
+			listener.onLightAdded(spot);
+		}
+	}
+	
+	private void removeBroadcast(DirectionalLight dir) {
+		for(EnvironmentListener listener : listeners) {
+			listener.onLightRemoved(dir);
+		}
+	}
+	
+	private void removeBroadcast(PointLight point) {
+		for(EnvironmentListener listener : listeners) {
+			listener.onLightRemoved(point);
+		}
+	}
+	
+	private void removeBroadcast(SpotLight spot) {
+		for(EnvironmentListener listener : listeners) {
+			listener.onLightRemoved(spot);
+		}
+	}

```

Comment:
Do we really need three different `addBroadcast` and `removeBroadcast` methods, for each light type? I think one for BaseLight should be sufficient.